Add .earcref to Archive Helper Class for Bulk Demo Exports

Rinual opened the pull request:
in theory adds earcref definitions to archivehelper.cs which should prevent Flagrum crashing when bulk exporting from demos.

Can't currently test to confirm while xmb1 is still broken, and changes were made to this class since I was last editing it, and my edits here were part of a larger project, but I believe this should do it, though you will likely know more than I with that. 

see https://github.com/Kizari/Flagrum/issues/192

Kizari reviewed (changes requested):
Continuing the discussion from #192, apologies this wasn't as straightforward as I expected, suppose the TODO at the top of that file is just as relevant as when it was first added!

You've added it to two maps. One is `RelativeExtensionMap` which is how Flagrum converts the source extensions (the ones at the end of URIs) to the proper file system extension that represents the actual file. Therefore, your first change is telling the asset explorer to export any files with `.earcref` at the end of the URI to a `.earc` file. Is this correct?

The second change affects the `RelativeExtension` list for FFXV, but not Forspoken. That's totally fine since we're looking to drop Forspoken support now anyway. I believe this list is used to populate the file type checkboxes in the bulk export modal. However, this is only meant to contain real file types, not those that are only present in URIs. As such, since `earcref` is being mapped back to `earc` (which is already present in this list), `earcref` should not be added to this list.

You also mentioned the presence of "@earcref" entries in #192. Assuming you meant `earcref@` entries, then this should be added to the `RelativeExtensionMap` as well, since we have done the same for `ebex@` and `prefab@`.

Once changes are made, I'll be sure to test it once I've fixed the other demo issues and merged this PR. Let me know if you have any further questions.

Rinual commented:
I’ll look into soon, but u believe i meant “ebex@.earcref” unless i missed something

Kizari commented:
Just noticed there's also `_uriToPathExtensionMap` in the `UriHelper` class. Would add `{"earcref", "earc"}` to this as well, since that's needed by the exporter IIRC.

Assuming it is `.ebex@.earcref` you were referring to as the other possibility, would also add `{"ebex@.earcref", "earc"}` to `_uriToPathExtensionMap` if it is representative of an `earc` file, and the same to `_map` under the `RelativeExtensionMap` class.

Rinual commented:
sorry, wasn't an *and*, it was the only iirc. I built an older build 1.5.14 which has xmb1 parsing back working, so I can confirm with the more minimalist changes you suggested.

Rinual commented:
mkay, that gets it *most of the way*, for exporting almost the entire demo
dx11.vfxbin is the newest exception

Rinual commented:
well, assuming its fine to add "dx11.vfxbin", to RelativeExtensions(LuminousGame type). That should be the last exception preventing bulk exporting from all game files.... though still should mention this is with 1.5.14 build

Kizari commented:
May as well add those extra keys for PD as well, yeah. They're pretty useless, but need to be able to bulk export the entire game successfully.

Rinual commented:
> May as well add those extra keys for PD as well, yeah. They're pretty useless, but need to be able to bulk export the entire game successfully.

Any advice based on where the file changes in the pr atm and why that last one i mentioned isnt working like the others?

Kizari commented:
Not really sure why that one in particular would be failing just based on the information here and in the PR. I'll do some testing on my end to see what's going wrong and get back to you.

Kizari commented:
Afraid I'm not sure what to make of `data://character/pr/pr73/model_000/materials/.getxfer.9796.1744.mega`. I don't run into this issue when exporting the demos, nor does that URI exist anywhere in PD, EpD, or retail on my end.

All of the other extensions you mentioned do exist on my end though, and they do resolve the issues now that I've fixed #192 in my local branch.

Kizari commented on src/Flagrum.Core/Archive/ArchiveHelper.cs:
```diff
@@ -255,8 +259,10 @@ public static string[] RelativeExtensions(LuminousGame type)
                 "dyn",
-                "earc",
+                "earc",	  
+                "ebex@.earcref",
```
This one isn't needed, since this list is only for the path extensions, not the URI extensions. Since `ebex@.earcref` maps to `earc`, just having `earc` in this list is sufficient.

Kizari commented on src/Flagrum.Core/Archive/ArchiveHelper.cs:
```diff
@@ -229,6 +232,7 @@ public static string[] RelativeExtensions(LuminousGame type)
         {
             return
             [
+                "1744.mega",
```
Still unsure where this extension came from, as it's not in my files at all for either demo or any retail build. Since you mentioned it was still erroring anyway, probably best to remove it. I don't mind if Flagrum does facilitate it if this isn't unique to you though, but would need to ensure it works before it'd be worth merging in.

Kizari commented on src/Flagrum.Core/Archive/ArchiveHelper.cs:
```diff
@@ -170,12 +170,15 @@ public partial class RelativeExtensionMap
         {"exr", "btex"},
         {"tga", "btex"},
         {"png", "btex"},
+        {"png.copy", "png"},
```
Everything in this dictionary looks good. Just a note though, the `_uriToPathExtensionMap` above is essentially the same as this map, except it covers all extensions, rather than only those where the value isn't the same as the key. As such, all of these new entries should be duplicated in that map.

Kizari reviewed (changes requested):
Have reviewed the file changes just to give you something more useful to work with since this conversation has ended up a bit longer than anticipated.

Since all the other fixes are done now, feel free to pull the upstream branch (i.e. `main` from `Kizari/Flagrum`) into your fork to test this PR on your end.

Rinual commented:
made those changes. 

however, though PD exported fine, including converting .exml to .XML

Exporting higher up folders just exports forever, no crashes, just keeps exporting with no files being added to the export directory. 

for some reason even just exporting the folder here `data://CUSA01709-app/menu/map/travel_menu/script/layout_travel_menu_files/menu/map/travel_menu/script`

is returning .exml even when .XML is selected, flagrum can read the .exml fine, and converts it .XML if exported manually.

Kizari commented:
> Exporting higher up folders for EPD just exports forever, no crashes, just keeps exporting with no files being added to the export directory.

Think this may have been caused by an oversight from years ago. The asynchronous export task was unobserved, so if any exceptions occurred during its operation, the task would fail silently with no feedback to the user (which would leave the loading indicator active). Commit cc218e557c8598fc2f2e505c3ed57613daf9a5b2 addresses this to prevent it happening again.

> for some reason even just exporting the folder here...

Judging by the URI, this is a loose file (of a type which is no longer indexed). You may have forgotten to reindex the Episode Duscae profile after changing to it, since you already reindexed Platinum Demo. If it's still there after reindexing, please let me know. **Make sure you reindex after pulling the latest changes from this repository's `main` branch and addressing the next point below in your PR, as it will cause issues if you don't index with all relevant file extensions present.**

---

As for why your export task was failing silently in the first place (I know it has given you errors before, but due to the nature of tasks, sometimes the exceptions will be observed even if the task is not, I think this has to do with the garbage collector) it's because we're missing more file extensions from EpD. See the variables below for which items they're missing, which should be added to your PR.

**_uriToPathExtensionMap**
```
{"elbin", "elbin"},
{"nav_ref", "nav_ref"},
{"orb.config", "orb.config"},
{"orb.fntbin", "orb.fntbin"},
{"style", "style"},
{"tmpdbin", "tmpdbin"},
{"wthcl", "wthclb"},
{"wthex", "wthexb"},
{"wthpe", "wthpeb"},
{"wthsky", "wthskyb"},
```

**_map** (in `RelativeExtensionMap` class)
```
{"wthcl", "wthclb"},
{"wthex", "wthexb"},
{"wthpe", "wthpeb"},
{"wthsky", "wthskyb"},
```

**RelativeExtensions** (first returned array in this function, same as where you've put the others)
```
"elbin",
"tmpdbin",
"wthclb",
"wthexb",
"wthpeb",
"wthskyb",
```

Rinual commented:
I think you are ignoring a few things:

1) it had it reindex and export fully in 1.5.14 with the changes above… multiple times, fully from root
2) i was deleting and re-indexing and having that issue only after your most recent changes

yes model previews used to work in 1.5.14, but there are still many unsupported demo btex types.

Rinual commented:
updated to the latest commit, added 
![image](https://github.com/user-attachments/assets/f176d64e-87a9-4038-a0e3-ffd7efe6a3fb)
as well as `data://data/ecology/bin/ecologyproj.elbin`
and `data://environment/world/heightmaps/bin/terrainmaterial.tmpdbin`

finished bulk export.

Rinual commented:
added in your recommended changes as well on top of the ones that were needed for the export.

Kizari commented on src/Flagrum.Core/Archive/ArchiveHelper.cs:
```diff
@@ -127,6 +137,10 @@ public class UriHelper : IUriHelper
+        {"wthcl", "wthcl"},
+        {"wthex", "wthex"},
+        {"wthpe", "wthpe"},
+        {"wthsky", "wthsky"},
```
These weather mappings aren't quite right, they are missing the `b` suffix on the right-hand side.

Rinual replied:
ah, yeah, i had them all added without the B at first in all sections, then noticed you had it changed elsewhere for the relative map, missed that you had it changed up here as well. should be done
